lib/Sema/CSSimplify.cpp

```diff
@@ -7683,10 +7683,24 @@ ConstraintSystem::simplifyKeyPathConstraint(
     return true;
   };
   
-  // We have a hole, the solver can't infer the key path type. So let's
-  // just assume this is solved.
-  if (shouldAttemptFixes() && keyPathTy->isHole()) {
-    return SolutionKind::Solved;
+  // If we have a hole somewhere in the key path, the solver won't be able to
+  // infer the key path type. So let's just assume this is solved.
+  if (shouldAttemptFixes()) {
+    if (keyPathTy->isHole())
+      return SolutionKind::Solved;
+
+    // If the root type has been bound to a hole, we cannot infer it.
+    if (getFixedTypeRecursive(rootTy, /*wantRValue*/ true)->isHole())
+      return SolutionKind::Solved;
+
+    // If we have e.g a missing member somewhere, a component type variable
+    // will have be marked as as a potential hole.
+    if (llvm::any_of(componentTypeVars, [&](TypeVariableType *tv) {
+          return tv->getImpl().getLocator()->isForKeyPathComponent() &&
+                 tv->getImpl().canBindToHole();
```

Comment:
I think it was sufficient originally to stop when for type variable to be marked as a hole but the meaning of this option has been changed to indicate that it could be a hole. Instead of just returning here we should wait with short-circuiting of key path application until at least some of the components have been _determined_ to be a hole. WDYT?

Reply:
@xedin That was my initial thought too, unfortunately types can currently back-propagate into the overload's type var, e.g for something like:

```
let kp: KeyPath<String, String> = \.bloop
```

Even though we record a fix for `bloop` being a missing member, and mark the type variable as being able to bind to a hole, the solver would rather bind it to `String` from the context. We probably need some way to mark the overload itself as missing rather than just not recording it in these cases, what do you think? 

Fortunately for now at least I believe overload type variables aren't marked as potential holes unless they're erroneous in some way, so I believe this logic should work for now without prematurely short-circuiting. I can drop in a FIXME to clean this up.

Reply:
That's actually correct behavior, we'd want to have as few holes as possible but it's unfortunate that it interferes with this. I think you should drop a TODO here and leave it for now.

We should re-design how everything works around key path. I have already mentioned that to @LucianoPAlmeida in one of his PRs that key path should work just like closures do today - resolve key path itself only when components are resolved, key path type inference is only attempted once the last type variable representing a component gets bound to a type.